runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,218 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
       return isNullTestImplicit;
    }
 
+
+/** \brief
+ *     Generates a dynamicCache test with helper call for instanceOf/ifInstanceOf node
+ *
+ *  \details
+ *     This funcition generates a sequence to check per site cache for object class and cast class before calling out to jitInstanceOf helper
+ */
+static
+void genInstanceOfDynamicCacheAndHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Register *castClassReg, TR::Register *objClassReg, TR::Register *resultReg, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *doneLabel, TR::LabelSymbol *helperCallLabel, TR::LabelSymbol *dynamicCacheTestLabel, TR::LabelSymbol *branchLabel, TR::LabelSymbol *trueLabel, TR::LabelSymbol *falseLabel, bool dynamicCastClass, bool generateDynamicCache, bool cacheCastClass, bool ifInstanceOf, bool trueFallThrough )
+   {
+   TR::Compilation                *comp = cg->comp();
+   bool needResult = resultReg != NULL;
+   if (!castClassReg)
+      castClassReg = cg->evaluate(node->getSecondChild());
+   int32_t maxOnsiteCacheSlots = comp->getOptions()->getMaxOnsiteCacheSlotForInstanceOf();
+   TR::Register *dynamicCacheReg = NULL;
+   int32_t addressSize = TR::Compiler->om.sizeofReferenceAddress();
+   /* Layout of the writable data snippet
+    * Case - 1 : Cast class is runtime variable
+    * [UpdateIndex][ObjClassSlot-0][CastClassSlot-0]...[ObjClassSlot-N][CastClassSlot-N]
+    * Case - 2 : Cast Class is interface / unresolved
+    * [UpdateIndex][ObjClassSlot-0]...[ObjClassSlot-N]
+    * If there is only one cache slot, we will not have header.
+    * Last bit of cached objectClass will set to 1 indicating false cast
+    */
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   if (generateDynamicCache)
+      {
+      TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
+      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
```

Comment:
I have checked function that estimates/emits snippet body. In the emit snippet body I see following loop,
Following functions is called from the OMR component,  (`omr/compiler/z/codegen/OMRCodeGenerator.cpp:8210`)
```
for (exp = self()->constantDataSnippetExponent(); exp > 0; exp--)
      {
      size = 1 << exp;
      first = true;
      for (auto writeableiterator = _writableList.begin(); writeableiterator != _writableList.end(); ++writeableiterator)
         {
       if (maxSize < (*writeableiterator)->getConstantSize())
            maxSize = size;

         if (HANDLE_CONSTANT_SNIPPET((*writeableiterator), size))
            {
            if (first)
               {
               first = false;
               self()->setBinaryBufferCursor((uint8_t *) (((uintptrj_t) (self()->getBinaryBufferCursor() + size - 1) / size) * size));
               }
            codeOffset = (*writeableiterator)->emitSnippetBody();
            if (codeOffset != NULL)
               {
               self()->setBinaryBufferCursor(codeOffset);
               }
            }
         }
      }
```
Where it will emit the snippet body only when `HANDLE_CONSTANT_SNIPPET((*writeableiterator), size)` returns true. Please note that `size` is defined as `1<<exp` at the beginning of loop. 
If you see definition of this function-like macro, 
```
   #define HANDLE_CONSTANT_SNIPPET(cursor, pow2Size) \
                      (cursor->getConstantSize() == pow2Size)
```
which returns true only if the snippet size is equal to pow2Size. That is why currently we only can have snippet size in power of two.  
@joransiu Are you aware of any historical reason behind this restriction?

Reply:
This seems silly. The constructor of `S390ConstantDataSnippet` has no mention of any of this. You would think the constructor would have asserted if there was an implicit assumption somewhere else in the code that we can only emit snippets of power of 2.

Thanks for explaining this. Can you please open up an issue to get this fixed and reference it here? Your code is fine.

Reply:
Opened up issue on `OMR` project as this part of code belongs to it. https://github.com/eclipse/omr/issues/1815